### DISCUSSION TITLE
Support GitHub Enterprise Server

### DISCUSSION THREAD

**Discussion by naikrovek** [lowlighter/metrics#1186] [Ideas]
I hate asking for this without also offering a PR, and I may still do that; I just wanted to ask politely for support to be added for the self-hosted GitHub Enterprise Server (GHES) appliance.

The APIs are the same, only the endpoint changes for GHES.  A single place to specify the root API endpoint would be all that is needed.

GHES v3 base endpoint looks like this: https://hostname.example.com/api/v3 
GHES v4 base endpoint looks like this: https://hostname.example.com/api/graphql

moving those out to a configuration variable somewhere would be 95% of the work required.  There are (probably) some differences for v4, and I don't use it enough to know.  There are some small differences for v3, but they are so small in number that I can't think of any.

Thank you.  As I said, if/when I get time, I will submit code to this effect on my own.

**Comment by lowlighter** [comment]
Hi 👋 !

It has been implemented with #1188 👍 

You can now use [`github_api_rest`](https://github.com/lowlighter/metrics/blob/master/source/plugins/core/README.md#github_api_rest) and [`github_api_graphql`](https://github.com/lowlighter/metrics/blob/master/source/plugins/core/README.md#github_api_graphql) to change both endpoints

Unfortunately I don't have GHES to test it out, but it should work since octokits already have a [`baseUrl`](https://octokit.github.io/rest.js/v18#usage) to support this
If you have some time to test it, don't hesitate to give some feedback on whether it works as expected or not 😄

**Reply by naikrovek** [reply]
Thank you.  I'll test this as soon as I get a chance.  I had no idea you would have this ready so quickly, so I committed my time to a bunch of other things.

**Reply by naikrovek** [reply]
This works once I omit the `/graphql` portion of the url from the URL I set with the `github_api_graphql`.  If I put the correct URL, the URL the plugins choose is `https://github.example.com/api/graphql/graphql` so if I use `https://github.example.com/api` then everything works.  That seems like it needs a fix.

None of the plugins I'm using seem to use the REST API or they aren't complaining with the full URL supplied.

**Reply by naikrovek** [reply]
also, sponsorship isn't a thing on GitHub Enterprise Server, so those data points can be omitted entirely on GHES and GitHub Enterprise Cloud (GHEC).  I didn't see a way to turn this off for the repositories or activity/community plugins.

**Reply by naikrovek** [reply]
By the way, even without those fixes, this is exactly what I wanted.  Thank you so much!

**Reply by lowlighter** [reply]
> This works once I omit the `/graphql` portion of the url from the URL I set with the `github_api_graphql`. If I put the correct URL, the URL the plugins choose is `https://github.example.com/api/graphql/graphql` so if I use `https://github.example.com/api` then everything works. That seems like it needs a fix.
> 
> None of the plugins I'm using seem to use the REST API or they aren't complaining with the full URL supplied.

I think the graphql octokit may actually auto-append the `/graphql` portion to the url passed. If that's the case I wonder if `github_api_graphql` option is actually useful and there are case where it would be different from `github_api`...

> also, sponsorship isn't a thing on GitHub Enterprise Server, so those data points can be omitted entirely on GHES and GitHub Enterprise Cloud (GHEC). I didn't see a way to turn this off for the repositories or activity/community plugins.

I think we could automatically hide the "unsupported options" of GHES (if there are others apart from sponsorship) for this case.

I don't want to add an option to specifically disable sponsorships else it'd mean that each stats should also have one too to be consistant, and this is probably better as a custom template or using `extras_css` to hide specific parts directly
